AES 256 CBC cipher

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->


- [x] `encrypt`, `decrypt` AES 256 CBC cipher impl
- [x] padding and unpadding impl
- [x] roundtrips for padding
- [x] roundtrips for encrypt/decrypt with and without padding enabled
- [x] golden generated from openssl for encrypt without salt
- [ ] 
->  Support for encryption with salt and golden generated from openssl to be tackled in the next PR

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number
adp-3327

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
add AES256CBC module

paweljakubas committed:
padding - part 1

paweljakubas committed:
add unpadding interface

paweljakubas committed:
add implementation to padding and unpadding

paweljakubas committed:
add padding roundtrips

more

paweljakubas committed:
introduce CipherMode and code refactoring using either

paweljakubas committed:
correct impl

paweljakubas committed:
add without padding encrypt/decrypt roundtrip

paweljakubas committed:
more tests and improve decrypt

paweljakubas committed:
first golden passing

paweljakubas committed:
more golden

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+            if len < 16 then
+                len
+            else
+                len `mod` 16
```
Hi @paweljakubas 

Just curious, why not just write:

```suggestion
        remaining = len `mod` 16
```

(Since for any natural number `n` in the range `[0, 15]`, `n mod 16 == n`)

paweljakubas replied:
yes, thanks

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+            if remaining == 0 then
+                B8.replicate 16 (toEnum 16)
+            else
+                B8.replicate (16 - remaining) (toEnum (16 - remaining))
```
And similarly:
```suggestion
        padding = B8.replicate (16 - remaining) (toEnum (16 - remaining))
```

Justification:
```hs
    B8.replicate (16 - remaining) (toEnum (16 - remaining))
 == B8.replicate (16 -         0) (toEnum (16 -         0)) -- remaining == 0
 == B8.replicate (16            ) (toEnum (16            )) -- x - 0 == x
 == B8.replicate  16              (toEnum  16             ) -- (x) == x
```

paweljakubas replied:
correct!

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+   CryptoFailed e -> Left e
+   CryptoPassed a -> Right a
+
+-- | Initialize IV
```
Just a suggestion (for ease-of-comprehension):
```suggestion
-- | Create an initialisation vector (IV).
```

paweljakubas replied:
:+1:

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+initIV
+    :: ByteString
+    -> Either CryptoError (IV AES256)
+initIV iv =
```
IMO, the juxtaposition of `init` and `IV` (an abbreviation of `initialisation vector`) seems a bit awkward. (As in "initialising the initialisation vector".)

Perhaps we could use a different verb here, such as `create`:
```suggestion
createIV iv =
```
(But this is just a nitpick!)

paweljakubas replied:
:+1:

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+initIV iv =
+    case makeIV iv of
+        Nothing -> Left CryptoError_IvSizeInvalid
+        Just ivOk -> pure ivOk
```
Not sure if you're fond of `maybeToEither`, but if you are, then we perhaps this could be simplified to:
```suggestion
initIV = maybeToEither (Left CryptoError_IvSizeInvalid) . makeIV
```

paweljakubas replied:
I am fan of either, maybe, ... combinators. Thanks for the suggestion

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+    -> ByteString
+    -- ^ secret key, needs to be 32 bytes
+    -> ByteString
+    -- ^ iv, needs to be 16 bytes
```
Given that this is a comment, perhaps it could be helpful to expand the abbreviation here:

```suggestion
    -- ^ initialisation vector (IV): must be 16 bytes
```

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+    in if len == 0 then
+           Nothing
+       else
+           Just $ BS.append payload padding
```
Using the suggestions above in conjunction with pattern guards, I think we can simplify this whole function to:

```suggestion
paddingPKCS7 payload
    | BS.null payload = Nothing
    | otherwise = Just $ BS.append payload padding
  where
    padding = B8.replicate paddingLength (toEnum paddingLength)
    paddingLength = 16 - (BS.length payload `mod` 16)
```

See https://github.com/cardano-foundation/cardano-wallet/pull/4518/commits

paweljakubas replied:
super! thanks!

jonathanknowles committed:
Wrap comments to satisfy line length limit.

jonathanknowles committed:
In `paddingPKCS7`: simplify definition of `remaining`.

jonathanknowles committed:
Move `AES256CBCSpec` to dedicated test suite for `crypto-primitives`.

As a result of this commit, the `crypto-primitives` has its own test
suite.

jonathanknowles committed:
In `paddingPKCS7`: simplify definition of `padding`.

Justification:

```
    B8.replicate (16 - remaining) (toEnum (16 - remaining))
 == B8.replicate (16 -         0) (toEnum (16 -         0)) -- remaining == 0
 == B8.replicate (16            ) (toEnum (16            )) -- x - 0 == x
 == B8.replicate  16              (toEnum  16             ) -- (x) == x
```

jonathanknowles committed:
In `paddingPKCS7`: use `where` clause instead of `let` and `in`.

jonathanknowles committed:
In `paddingPKCS7`: use `BS.null` instead of `BS.length == 0`.

jonathanknowles committed:
In `paddingPKCS7`: use pattern guards.

jonathanknowles committed:
In `paddingPKCS7`: simplify definition of `remaining`.

jonathanknowles committed:
In `paddingPKCS7`: simplify definition of `padding`.

jonathanknowles commented on lib/crypto-primitives/crypto-primitives.cabal:
```diff
@@ -25,6 +25,7 @@ library
 
   hs-source-dirs:     src
   exposed-modules:    Cryptography.Cipher.ChaChaPoly1305
+                      Cryptography.Cipher.AES256CBC
```
I just noticed that the test suite for this is in another package (`unit.cabal`).

I propose making a dedicated test suite for `crypto-primitives`, what do you think?

I made a commit that does this here: 2c1a37045e4c177656c4b044e2e0b0152dc8c635

Feel free to cherry-pick!

jonathanknowles reviewed (commented):
Hi @paweljakubas 

I've just taken a quick look, and I have some suggestions.

I put some of the suggestions in this PR: https://github.com/cardano-foundation/cardano-wallet/pull/4518. Feel free to cherry-pick if you like!

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -0,0 +1,169 @@
+            if (fromEnum lastByte) >= BS.length payload then
+                payload
+            else
+                BS.dropEnd (fromEnum lastByte) payload
```
Hi @paweljakubas 

Just curious -- why is the first branch of the `if` required? According to the [spec](https://www.ibm.com/docs/en/zos/2.4.0?topic=rules-pkcs-padding-method), couldn't we replace this with:

```suggestion
                let paddingLength = fromEnum lastByte in           
                BS.dropEnd paddingLength payload
```

With this change, the test suite still passes.

(Apologies if I have missed something from the spec.)

paweljakubas replied:
no no, everything passes. Meaning simplification you applied is correct!

paweljakubas committed:
clean unit test cabal

paweljakubas committed:
mv KDF testing to crypto-primitives's test hierarchy

paweljakubas committed:
hlint

paweljakubas committed:
more suggestion applied

jonathanknowles committed:
Minor formatting fixes.

Wrap lines to fit within line length limit (80 chars).
Use 4-space indents where possible.

jonathanknowles committed:
In `unpaddingPKCS7`, move `cut` to `where` clause.

jonathanknowles committed:
In `unpaddingPKCS7`, inline definition of `initLast`.

jonathanknowles committed:
In `unpaddingPKCS7`, simplify definition of `cut`.

jonathanknowles committed:
In `unpaddingPKCS7`, extract out `paddingLength`.

This arguably makes the code more self-documenting, as we are explicitly
documenting the fact that it's the padding that we're removing, which
has a defined length.

jonathanknowles committed:
In `unpaddingPKCS7`, rename function `cut` to `stripPadding`.

jonathanknowles committed:
Rename function `paddingPKCS7` to `padPKCS7`.

This changes the name of the function from a noun to a verb, which is
arguably more idiomatic.

jonathanknowles committed:
Rename `unpaddingPKCS7` to `unpadPKCS7`.

This changes the name of the function from a noun to a verb, which is
arguably more idiomatic.

jonathanknowles committed:
Minor tweaks to documentation.

jonathanknowles committed:
Use more explicit imports within `AES256CBC`.

This makes it easier for readers to establish the provenance of imported
symbols, instead of having to hunt around.

jonathanknowles committed:
More suggestions for AES 256 CBC cipher implementation (PR #4517) (#4523)

This PR contains more suggestions for PR #4517.

jonathanknowles reviewed (approved):
Hi @paweljakubas!

I only have one suggestion (at the moment), which is to adjust `unpadPKCS7` so that it fails if the padding does not meet the specification.

However, given that this is a work-in-progress, I think we can do that in a future PR.
